demos-go/examples/ecdsa-mpc-with-backup/main.go

```diff
@@ -3,20 +3,204 @@ package main
+	pub, err := x509.ParsePKCS1PublicKey(ek)
+	if err != nil {
+		return nil, nil, err
+	}
+	if pub.Size() != 256 {
+		return nil, nil, fmt.Errorf("invalid RSA modulus size: got %d bits", pub.Size()*8)
+	}
+	if pub.E != 65537 {
+		return nil, nil, fmt.Errorf("unsupported RSA public exponent")
+	}
+
+	// --- Derive independent materials from rho, bound to key N and DS ---
+	salt := pub.N.Bytes() // binds derivations to recipient key
+
+	// Derive the OAEP seed (exactly Hash.Size() bytes for SHA-256)
+	var oaepSeed [32]byte
+	if _, err := io.ReadFull(hkdf.New(sha256.New, rho[:], salt, []byte(kemDS+"|oaep-seed")), oaepSeed[:]); err != nil {
+		return nil, nil, fmt.Errorf("hkdf: %w", err)
+	}
+
+	// Derive the shared secret independently from the OAEP seed
+	ss := make([]byte, 32)
+	if _, err := io.ReadFull(hkdf.New(sha256.New, rho[:], salt, []byte(kemDS+"|ss")), ss); err != nil {
+		zero(oaepSeed[:])
+		return nil, nil, fmt.Errorf("hkdf: %w", err)
+	}
+
+	// Deterministic OAEP randomness: rsa.EncryptOAEP reads exactly Hash.Size() bytes.
+	r := bytes.NewReader(oaepSeed[:])
+	ct, err := rsa.EncryptOAEP(sha256.New(), r, pub, ss, []byte(kemLabel))
+	zero(oaepSeed[:]) // best-effort wipe
+	if err != nil {
+		zero(ss)
+		return nil, nil, err
+	}
+
+	// Deterministic: (pub, rho) -> (ct, ss)
+	return ct, ss, nil
+}
+
+func (rsaDemoKEM) Decapsulate(skHandle unsafe.Pointer, ct []byte) ([]byte, error) {
+	// Expect cmem_t pointing to private key bytes
+	type cmem_t struct {
+		data *byte
+		size int32
+	}
+	cm := (*cmem_t)(skHandle)
+	if cm == nil || cm.data == nil || cm.size <= 0 {
+		return nil, fmt.Errorf("kem: decapsulation failed") // uniform error
+	}
+	// Sanity cap to avoid DoS; PKCS#1 DER for 2048-bit is ~1–2KB
+	if cm.size < 256 || cm.size > 8192 {
+		return nil, fmt.Errorf("kem: decapsulation failed")
+	}
+
+	// Copy foreign memory into Go memory before parsing (safer if caller frees it)
+	dk := unsafe.Slice((*byte)(unsafe.Pointer(cm.data)), int(cm.size))
+	dkCopy := make([]byte, len(dk))
+	copy(dkCopy, dk)
+	// Ensure cm isn’t GC’d early
+	runtime.KeepAlive(cm)
+
+	prv, err := x509.ParsePKCS1PrivateKey(dkCopy)
+	zero(dkCopy) // best-effort wipe of private key bytes copy
+	if err != nil {
+		return nil, fmt.Errorf("kem: decapsulation failed")
+	}
+	// Pin modulus size to 2048
+	if prv.Size() != 256 {
+		return nil, fmt.Errorf("kem: decapsulation failed")
+	}
+	// Basic sanity on ciphertext length
+	if len(ct) != prv.Size() {
+		return nil, fmt.Errorf("kem: decapsulation failed")
+	}
+
+	// Decrypt with blinding; output still deterministic wrt inputs
+	ss, err := rsa.DecryptOAEP(sha256.New(), rand.Reader, prv, ct, []byte(kemLabel))
+	if err != nil {
+		return nil, fmt.Errorf("kem: decapsulation failed")
+	}
+	if len(ss) != 32 {
+		// Should not occur if encaps follows the contract, but guard anyway
+		return nil, fmt.Errorf("kem: decapsulation failed")
+	}
+
+	out := make([]byte, 32)
+	copy(out, ss)
+	zero(ss) // wipe temp
```

Comment:
this is what I was referring to above (using secure wipe instead of local zero function)